aptos-move/block-executor/src/executor_utilities.rs

```diff
@@ -0,0 +1,274 @@
+                    (resource_tag, bytes)
+                })
+                .collect();
+
+            // TODO[agg_v2](fix): Handle potential serialization failures.
+            bcs::to_bytes(&btree)
+                .map_err(|e| {
+                    resource_group_error(format!("Unexpected resource group error {:?}", e))
+                })
+                .map(|group_bytes| {
+                    metadata_op.set_bytes(group_bytes.into());
+                    (group_key, metadata_op)
+                })
+        })
+        .collect()
+}
+
+pub(crate) fn gen_id_start_value(sequential: bool) -> u32 {
+    // IDs are ephemeral. Pick a random prefix, and different each time,
+    // in case exchange is mistakenly not performed - to more easily catch it.
+    // And in a bad case where it happens in prod, to and make sure incorrect
+    // block doesn't get committed, but chain halts.
+    // (take a different range from parallel execution, to even more easily differentiate)
+
+    let offset = if sequential { 0 } else { 1000 };
+    thread_rng().gen_range(1 + offset, 1000 + offset) * 1_000_000
+}
+
+pub(crate) fn map_id_to_values_in_group_writes<
+    T: Transaction,
+    S: TStateView<Key = T::Key> + Sync,
+    X: Executable + 'static,
+>(
+    finalized_groups: Vec<(T::Key, T::Value, Vec<(T::Tag, ValueWithLayout<T::Value>)>)>,
+    latest_view: &LatestView<T, S, X>,
+) -> ::std::result::Result<Vec<(T::Key, T::Value, Vec<(T::Tag, Arc<T::Value>)>)>, PanicError> {
+    let mut patched_finalized_groups = Vec::with_capacity(finalized_groups.len());
+    for (group_key, group_metadata_op, resource_vec) in finalized_groups.into_iter() {
+        let mut patched_resource_vec = Vec::with_capacity(resource_vec.len());
+        for (tag, value_with_layout) in resource_vec.into_iter() {
+            let value = match value_with_layout {
+                ValueWithLayout::RawFromStorage(value) => value,
+                ValueWithLayout::Exchanged(value, None) => value,
+                ValueWithLayout::Exchanged(value, Some(layout)) => Arc::new(
+                    replace_ids_with_values(&value, layout.as_ref(), latest_view)?,
+                ),
+            };
+            patched_resource_vec.push((tag, value));
+        }
+        patched_finalized_groups.push((group_key, group_metadata_op, patched_resource_vec));
+    }
+    Ok(patched_finalized_groups)
+}
+
+// For each delayed field in resource write set, replace the identifiers with values
+// (ignoring other writes). Currently also checks the keys are unique.
+pub(crate) fn map_id_to_values_in_write_set<
+    T: Transaction,
+    S: TStateView<Key = T::Key> + Sync,
+    X: Executable + 'static,
+>(
+    resource_write_set: Vec<(T::Key, Arc<T::Value>, Arc<MoveTypeLayout>)>,
+    latest_view: &LatestView<T, S, X>,
+) -> ::std::result::Result<Vec<(T::Key, T::Value)>, PanicError> {
+    // Note: can remove or debug-only & be less defensive in the future / check elsewhere.
```

Comment:
hm, that would not check duplication between WriteWithDelayedFields and InPlaceDelayedFieldChange? The resource write set here is a combination of those two. I guess those are supposed to be disjoint by construction? (but it is an invariant of input in a way, although I agree this check feels like too much trouble)

Reply:
VMChangeSet.resource_write_set contains all different variants of AbstractResourceWriteOp together.

when patching we go one by one, and modify from some special variant into WriteOp variant.

if you have both WriteWithDelayedFields and InPlaceDelayedFieldChange, first will modify it, second will fail because it is now WriteOp, which is not allowed to be patched.

Reply:
purged :)